app/models/git_repository.rb

```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
     raise MiqException::Error, err.message
   end
 
+  def git_lock_filename
+    @git_lock_filename ||= LOCKFILE_DIR.join(id.to_s)
+  end
+
+  def acquire_git_lock
+    return false if defined?(@git_lock) && !@git_lock.nil?
+
+    FileUtils.mkdir_p(LOCKFILE_DIR)
+
+    @git_lock = File.open(git_lock_filename, File::RDWR | File::CREAT, 0o644)
+    @git_lock.flock(File::LOCK_EX)                         # block waiting for lock
+    @git_lock.write("#{Process.pid} - #{Time.zone.now}\n") # for debugging
+    @git_lock.flush                                        # write current data
+    @git_lock.truncate(@git_lock.pos)                      # clean up remaining chars
+
+    true
+  end
+
+  def release_git_lock
+    return if !defined?(@git_lock) || @git_lock.nil?
+
+    @git_lock.flock(File::LOCK_UN)
+    @git_lock.close
+
+    # TODO:  unlink?  Unsure how to handle this if other processes are waiting
+    # on this though...
```

Comment:
I don't think we need to unlock because the next process is going to reopen the file.

Reply:
unlock or unlink @Fryguy ?

Reply:
> I don't think we need to unlock because the next process is going to reopen the file.

Granted, not an expert on `File.flock`, but all of the files are "waiting on" or "acquiring" the lock AFTER they have opened the file.  So my thought here was that we would be pulling the rug from under the other processes that are waiting at their own `@git_lock.flock(File::LOCK_EX)` calls for the one with the lock to finish, and if the process with the lock deletes the file when releasing, I am pretty sure that would cause an IO error.

Reply:
Yeah, this was one of the reasons I was suggesting we just leave the file around. I don't think we ever really have to remove is unless the repo is deleted.

Reply:
> unlock or unlink @Fryguy ?

Typo fail...yes, unlink

Reply:
Agreed...don't unlink the file for all of the reasons mentioned.